code/modules/xenoarcheaology/effect.dm

```diff
@@ -3,19 +3,45 @@
 	var/effect = EFFECT_TOUCH
 	var/effectrange = 4
 	var/trigger = TRIGGER_TOUCH
-	var/atom/holder
+	var/datum/component/artifact_master/master
 	var/activated = 0
-	var/chargelevel = 0
+	var/chargelevel = 1
 	var/chargelevelmax = 10
 	var/artifact_id = ""
 	var/effect_type = 0
 
-/datum/artifact_effect/New(var/atom/location)
+	var/req_type = /atom/movable
+
+	var/image/active_effect
+	var/effect_icon = 'icons/effects/effects.dmi'
+	var/effect_state = "sparkles"
+	var/effect_color = "#ffffff"
+
+// The last time the effect was toggled.
+	var/last_activation = 0
+
+/datum/artifact_effect/Destroy()
+	if(master)
+		master = null
+
 	..()
-	holder = location
+
+/datum/artifact_effect/New(var/datum/component/artifact_master/newmaster)
+	..()
+	if(!newmaster)
+		qdel(src)
+		return
+
+	master = newmaster
 	effect = rand(0, MAX_EFFECT)
 	trigger = rand(0, MAX_TRIGGER)
 
+	if(effect_icon && effect_state)
+		if(effect_state == "sparkles")
+			effect_state = "sparkles_[rand(1,4)]"
+		active_effect = image(effect_icon, null, effect_state)
```

Comment:
Second arg for image() is state isn't it?

Reply:
If it is, that's strange. VSCode lookup said it's loc

Reply:
Might be magic BYOND horseshit, other usage of image() in polaris goes `image(icon, icon_state)`

Reply:
Yup, it's BYOND esoterica ([docs](https://www.byond.com/docs/ref/#/proc/image):
The official invocation is
`    image(icon,loc,icon_state,layer,dir) `
but there's a special case where the second argument is a string:

> Another common use of images is in making an overlay:
> ```
> overlays += image('pants.dmi',icon_state = "red")
> ```
> Since the loc argument could never be a text string, the above statement can be further shortened:
> ```
> overlays += image('pants.dmi',"red")
> ```